src/Microsoft.AspNet.Mvc.Xml/RequiredValidationHelper.cs

```diff
@@ -0,0 +1,97 @@
+{
+    public static class RequiredValidationHelper
+    {
+        public static void Validate(Type modelType, ModelStateDictionary modelStateDictionary)
+        {
+            Validate(modelType, modelStateDictionary, new HashSet<Type>());
+        }
+
+        private static void Validate(
+            Type modelType,
+            ModelStateDictionary modelStateDictionary,
+            HashSet<Type> visited)
+        {
+            if (modelType.IsGenericType())
+            {
+                var enumerableOfT = modelType.ExtractGenericInterface(typeof(IEnumerable<>));
+                if (enumerableOfT != null)
+                {
+                    modelType = enumerableOfT.GetGenericArguments()[0];
+                }
+            }
+
+            if (ExcludeTypeFromValidation(modelType))
+            {
+                return;
+            }
+
+            // Avoid infinite loop in case of self-referencing properties
+            if (visited.Contains(modelType))
+            {
+                return;
+            }
+
+            var referenceTypeProperties = new List<PropertyInfo>();
+            foreach (var property in modelType.GetProperties(BindingFlags.Public | BindingFlags.Instance))
+            {
+                if (property.PropertyType.IsValueType() && !property.PropertyType.IsNullableValueType())
+                {
+                    var required = property.GetCustomAttribute(typeof(RequiredAttribute), inherit: true);
+                    if (required == null)
+                    {
+                        continue;
+                    }
+
+                    var hasDataMemberRequired = false;
+
+                    var dataMemberRequired = (DataMemberAttribute)property.GetCustomAttribute(
+                        typeof(DataMemberAttribute),
+                        inherit: true);
+
+                    if (dataMemberRequired != null && dataMemberRequired.IsRequired)
+                    {
+                        hasDataMemberRequired = true;
+                    }
+
+                    if (!hasDataMemberRequired)
+                    {
+                        var propertyType = property.DeclaringType;
+                        modelStateDictionary.TryAddModelError(
+                            propertyType.FullName,
+                            Resources.FormatRequiredProperty_MustHaveDataMemberRequired(
+                                property.Name,
+                                propertyType.FullName));
+                    }
+                }
+                else
+                {
+                    referenceTypeProperties.Add(property);
+                }
+            }
+
+            visited.Add(modelType);
+
+            foreach (var referenceTypeProperty in referenceTypeProperties)
+            {
+                Validate(referenceTypeProperty.PropertyType, modelStateDictionary, visited);
+            }
+        }
+
+        private static bool ExcludeTypeFromValidation(Type modelType)
```

Comment:
Just like in DefaultObjectValidator, even here we ideally do not want to validate non-user defined types...i am not sure if i should depend on the type exclude filters in the formatter...